packages/flutter/test/material/navigation_rail_test.dart

```diff
@@ -0,0 +1,2150 @@
+      label: Text('Abc'),
+    ),
+    NavigationRailDestination(
+      icon: Icon(Icons.bookmark_border),
+      selectedIcon: Icon(Icons.bookmark),
+      label: Text('Def'),
+    ),
+    NavigationRailDestination(
+      icon: Icon(Icons.star_border),
+      selectedIcon: Icon(Icons.star),
+      label: Text('Ghi'),
+    ),
+    NavigationRailDestination(
+      icon: Icon(Icons.hotel),
+      selectedIcon: Icon(Icons.home),
+      label: Text('Jkl'),
+    ),
+  ];
+}
+
+Future<void> _pumpNavigationRail(
+  WidgetTester tester, {
+  double textScaleFactor = 1.0,
+  NavigationRail navigationRail,
+}) async {
+  await tester.pumpWidget(
+    MaterialApp(
+      home: Builder(
+        builder: (BuildContext context) {
+          return MediaQuery(
+            data: MediaQuery.of(context).copyWith(textScaleFactor: textScaleFactor),
+            child: Scaffold(
+              body: Row(
+                children: <Widget>[
+                  navigationRail,
+                  const Expanded(
+                    child: Text('body'),
+                  ),
+                ],
+              ),
+            ),
+          );
+        },
+      ),
+    ),
+  );
+}
+
+Future<void> _pumpLocalizedTestRail(WidgetTester tester, { NavigationRailLabelType labelType, bool extended = false }) async {
+  await tester.pumpWidget(
+    Localizations(
+      locale: const Locale('en', 'US'),
+      delegates: const <LocalizationsDelegate<dynamic>>[
+        DefaultMaterialLocalizations.delegate,
+        DefaultWidgetsLocalizations.delegate,
+      ],
+      child: MaterialApp(
+        home: Scaffold(
+          body: Row(
+            children: <Widget>[
+              NavigationRail(
+                selectedIndex: 0,
+                extended: extended,
+                destinations: _destinations(),
+                labelType: labelType,
+              ),
+              const Expanded(
+                child: Text('body'),
+              ),
+            ],
+          ),
+        ),
+      ),
+    ),
+  );
+}
+
+RenderBox _iconRenderBox(WidgetTester tester, IconData iconData) {
+  return tester.firstRenderObject<RenderBox>(
+    find.descendant(
+      of: find.byIcon(iconData),
+      matching: find.byType(RichText),
+    ),
+  );
+}
+
+RenderBox _labelRenderBox(WidgetTester tester, String text) {
+  return tester.firstRenderObject<RenderBox>(
+    find.descendant(
+      of: find.text(text),
+      matching: find.byType(RichText),
+    ),
+  );
+}
+
+TextStyle _iconStyle(WidgetTester tester, IconData icon) {
+  return tester.widget<RichText>(
+    find.descendant(
+      of: find.byIcon(icon),
+      matching: find.byType(RichText),
+    ),
+  ).text.style;
+}
+
+Finder _opacityAboveLabel(String text) {
+  return find.ancestor(
+    of: find.text(text),
+    matching: find.byType(Opacity),
+  );
+}
+
+// Only valid when labelType != all.
+double _labelOpacity(WidgetTester tester, String text) {
+  final Opacity opacityWidget = tester.firstWidget<Opacity>(
```

Comment:
The use of `firstWidget` here and elsewhere should be reconsidered. If we really only expect to find one widget, then `tester.widget`. Using `firstWidget` implies that we believe there may be more than one.

Reply:
Removed from here and elsewhere, and added a comment where it is necessary (such as the _railMaterial)